hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java

```diff
@@ -309,6 +310,16 @@ public WALEntryBatch poll(long timeout) throws InterruptedException {
     return entryBatchQueue.poll(timeout, TimeUnit.MILLISECONDS);
   }
 
+  public void clearWALEntryBatch() {
```

Comment:
Should this be atomic? Is there anything to prevent two racing threads: one calling this `clearWALEntryBatch()` method and another trying to normally consume the next `WALEntryBatch`?

I think this could lead to a double-decrement.

Reply:
We are only calling it from `ReplicationSource.terminate` (see line #606), after we certify that neither the shipper, nor the reader threads are alive anymore, so I don't think it would be an issue. Of course, there's the risk someone inadvertently call this method somewhere else, so maybe we should put a warning comment? I don't think there's any gain of synchronising accesses to totalBufferUsed variable here, concurrent threads could still succeed on the double decrement, if we call clearWALEntryBatch while shipper thread is still running.

Reply:
> We are only calling it from ReplicationSource.terminate (see line #606), after we certify that neither the shipper, nor the reader threads are alive anymore

Actually, we could move that check to clearWALEntryBatch method itself, and since shipper has a reference to reader, but not the other way around, we can move it to shipper, instead? Let me give it a try.

Reply:
> after we certify that neither the shipper, nor the reader threads are alive anymore, so I don't think it would be an issue. Of course, there's the risk someone inadvertently call this method somewhere else, so maybe we should put a warning comment?

Also OK to just put a warning if moving this check doesn't make things more clear :). Thanks for clarifying for me.